### PR TITLE
[DirectX] Fix DXIL part header version encoding

### LINKED ISSUE #89952: [DirectX] Fix DXIL part header version encoding
Today we are encoding the DXIL version in the DXIL part's program header backwards. Compiling a shader targeting `lib_6_8` with DXC then disassembling it with `obj2yaml` prints:

```
      DXILMajorVersion: 8
      DXILMinorVersion: 1
```

This probably means we're crossing the ordering of the fields somewhere in our handling.

We also need to set the DXIL version from the triple sub-architecture. Right now during DXIL header generation we are not setting the version. See [here](https://github.com/llvm/llvm-project/blob/main/llvm/lib/MC/MCDXContainerWriter.cpp#L121) where we set the shader model version but we don't set the DXIL version.

## Acceptance Criteria

* Updated obj2yaml tests verifying correct reproduction of major and minor versions.
* Updated llc -> obj2yaml testing verifying correct code generation from llc.

### PR DESCRIPTION
**Pull request by python3kgae**
Move MinorVersion be the lower 8 bit.
Set DXIL version in DXContainerObjectWriter::writeObject.

Fixes #89952

### CODE REVIEW

**Review by llvm-beanz — APPROVED** [review]

File: llvm/tools/llc/llc.cpp [hunk 1]
```diff
@@ -537,8 +537,8 @@ static int compileModule(char **argv, LLVMContext &Context) {
       // If we are supposed to override the target triple, do so now.
       std::string IRTargetTriple = DataLayoutTargetTriple.str();
       if (!TargetTriple.empty())
-        IRTargetTriple = Triple::normalize(TargetTriple);
-      TheTriple = Triple(IRTargetTriple);
+        IRTargetTriple = TargetTriple;
+      TheTriple = Triple(Triple::normalize(IRTargetTriple));
```
Comment (damyanp) [inline llvm/tools/llc/llc.cpp:541]: Doesn't `Triple::normalize` return a `Triple`, so it doesn't need to be wrapped inside `Triple()`?

**Review by damyanp — APPROVED** [review]

**Review by bharadwajy — COMMENTED** [review]
I see failures of LLVM regression tests (target `check-llvm`) with these changes.

Total Discovered Tests: 59475
  Skipped          :   346 (0.58%)
  Unsupported      : 27503 (46.24%)
  Passed           : 31174 (52.42%)
  Expectedly Failed:    63 (0.11%)
  Failed           :   389 (0.65%)
FAILED: test/CMakeFiles/check-llvm /workspace/github/oss/llvm-build/hlsl/Debug/test/CMakeFiles/check-llvm 

No such failures without this change (i.e., commit set to 8d9b15497d70ac782d7d01a2d606f9fec7e7f642) 

Total Discovered Tests: 59474
  Skipped          :   346 (0.58%)
  Unsupported      : 27503 (46.24%)
  Passed           : 31562 (53.07%)
  Expectedly Failed:    63 (0.11%)

Failures appear to be related to changes in `llvm/tools/llc/llc.cpp`

Please double check. Thanks!

**Comment by python3kgae** [comment]
> I see failures of LLVM regression tests (target `check-llvm`) with these changes.
> 
> Total Discovered Tests: 59475 Skipped : 346 (0.58%) Unsupported : 27503 (46.24%) Passed : 31174 (52.42%) Expectedly Failed: 63 (0.11%) Failed : 389 (0.65%) FAILED: test/CMakeFiles/check-llvm /workspace/github/oss/llvm-build/hlsl/Debug/test/CMakeFiles/check-llvm
> 
> No such failures without this change (i.e., commit set to [8d9b154](https://github.com/llvm/llvm-project/commit/8d9b15497d70ac782d7d01a2d606f9fec7e7f642))
> 
> Total Discovered Tests: 59474 Skipped : 346 (0.58%) Unsupported : 27503 (46.24%) Passed : 31562 (53.07%) Expectedly Failed: 63 (0.11%)
> 
> Failures appear to be related to changes in `llvm/tools/llc/llc.cpp`
> 
> Please double check. Thanks!

Thanks. Looking at it.

File: llvm/tools/llc/llc.cpp [hunk 2]
```diff
@@ -537,7 +537,9 @@ static int compileModule(char **argv, LLVMContext &Context) {
       // If we are supposed to override the target triple, do so now.
       std::string IRTargetTriple = DataLayoutTargetTriple.str();
       if (!TargetTriple.empty())
-        IRTargetTriple = Triple::normalize(TargetTriple);
+        IRTargetTriple = TargetTriple;
+      if (!IRTargetTriple.empty())
+        IRTargetTriple = Triple::normalize(IRTargetTriple);
```
Comment (bogner) [inline llvm/tools/llc/llc.cpp:542]: I don't think we should be changing `llc` to choose to always normalize triples buried in a DirectX change like this. This is a pretty significant change to the behaviour of llc when a module has an non-normalized triple, and if we really want to normalize the triple unconditionally I think that's a bigger change that needs wider discussion. Note that the Triple class deliberately does not normalize on construction, so that non-normalized triples can be worked with when appropriate (such as if it's intentionally broken for testing purposes, or something along those lines)

Comment (python3kgae) [reply llvm/tools/llc/llc.cpp]: Updated.

File: llvm/tools/llc/llc.cpp [hunk 3]
```diff
@@ -537,9 +537,7 @@ static int compileModule(char **argv, LLVMContext &Context) {
       // If we are supposed to override the target triple, do so now.
       std::string IRTargetTriple = DataLayoutTargetTriple.str();
       if (!TargetTriple.empty())
-        IRTargetTriple = TargetTriple;
-      if (!IRTargetTriple.empty())
-        IRTargetTriple = Triple::normalize(IRTargetTriple);
+        IRTargetTriple = Triple::normalize(TargetTriple);
```
Comment (bharadwajy) [inline llvm/tools/llc/llc.cpp:540]: Is this change still needed? Is it possible to avoid making changes to this file?

Comment (python3kgae) [reply llvm/tools/llc/llc.cpp:540]: No. There's no change in llc.cpp now.

**Review by bharadwajy — APPROVED** [review]
Just a reminder to make sure the sentence 

"Also always normalize target triple in llc to get DXIL version from subArch." 

is edited out of the "Squash and Merge" commit message.

File: llvm/lib/TargetParser/Triple.cpp [hunk 4]
```diff
@@ -1510,6 +1510,8 @@ VersionTuple Triple::getDXILVersion() const {
   if (getArch() != dxil || getOS() != ShaderModel)
     llvm_unreachable("invalid DXIL triple");
   StringRef Arch = getArchName();
+  if (Arch == "dxil")
```
Comment (bogner) [inline llvm/lib/TargetParser/Triple.cpp:1513]: I think you can say `getSubArch() == NoSubArch` here instead of using string comparison

**Review by bogner — APPROVED** [review]

File: llvm/lib/TargetParser/Triple.cpp (hunk 4 above)
Comment (python3kgae) [reply llvm/lib/TargetParser/Triple.cpp]: Fixed.

**Comment by python3kgae** [comment]
> Just a reminder to make sure the sentence
> 
> "Also always normalize target triple in llc to get DXIL version from subArch."
> 
> is edited out of the "Squash and Merge" commit message.
Good catch.
Updated.
